Add article translation feature

tr4nt0r opened the pull request:
Adds a method to get article translations and load article translation table if it is not already downloaded. Methods for sending and receiving articles from and to the Bring API have been extended to translate article names if a locale has been provided as argument. Also adds a method to get user settings, which includes also language settings for each list.

tr4nt0r committed:
Add article translation feature

Adds a method to get article translations and load article translation
table if it is not already downloaded. Methods for sending and receiving
articles from and to the Bring API have been extended to translate
article names if a locale has been provided as argument.
Also adds a method to get user settings, which includes also language
settings for each list.

tr4nt0r committed:
Merge branch 'main' into article_translations

tr4nt0r committed:
Rewrite getUserSettings and get ArticleTranslations to async

tr4nt0r commented:
This is also updated now to use aiohttp and conflicts were resolved.

I think this is pretty important to have in the library as Bring has a pretty extensive catalog of default articles. When selecting from one of these articles, the app internally translates them to the default language (which funnily enough is swiss german) before submitting it to the API. So if on the app, you put "Chicken breast" on the list, the API will return "Pouletbrüstli". This has then to be translated back into the lists language.

tr4nt0r committed:
fixed typo

miaucl commented on src/python_bring_api/bring.py:
```diff
@@ -713,3 +736,155 @@ async def notifyAsync(self, listUuid: str, notificationType: BringNotificationTy
+        BringParseException
+            If the parsing of the request response fails.
+       """ 
+        _supported_locales = {
```
This would be good to have in separate file and refer to it from the README.

tr4nt0r replied:
moved it for now to the __init__ but maybe to cleanup we can move some stuff to a const.py

miaucl commented on src/python_bring_api/bring.py:
```diff
@@ -713,3 +736,155 @@ async def notifyAsync(self, listUuid: str, notificationType: BringNotificationTy
+            }
+        if locale not in _supported_locales:
+            _LOGGER.debug('Locale %s not supported by Bring. Ommitting translation.', locale)
+            return {}
```
This is an error imo, and should raise.

miaucl replied:
Might even have to be factored out to a "check helper method" and called from each function to verify that the local provided is valid (or `None`).

tr4nt0r replied:
or should we just not check and if the request returns a 404 we know anyway that the language is not supported.

miaucl replied:
Might be a good idea as well!

tr4nt0r replied:
The webserver does not return a 404, it returns a 200 and loads the webApp which results in a JSONDecoderError 🙄

miaucl replied:
Haha, wtf

miaucl commented:
Overall, it would also be great to put some words about translation into the README anyway.

tr4nt0r committed:
add raise when unsupported locale and fix indentation

tr4nt0r committed:
Update README.md

miaucl reviewed (approved):
Should be good to be usable in HA :)

tr4nt0r committed:
fix bug in item translation loop

bring api returns different json structure when x-bring-client android is used.
The webApp client is considered legacy mode by the bring api.

eliasball commented:
I am terribly sorry that it is taking me so long to review all the PRs, there is a lot of other stuff that I have to take care of right now. How urgently do you need these changes reviewed and merged?

tr4nt0r commented:
not that urgent anymore, we forked your library because we wanted to go in a slightly different direction for some things. 

https://github.com/miaucl/bring-api

eliasball commented:
I see, that makes sense. Great job with your fork, your implementation looks really good! Good luck with everything :)

eliasball commented:
A little side note: I see that you have left my license/copyright notice in, which is good and also actually required by the MIT license. However, you might want to/should add your own new copyright notice/license e.g. in the same file above the old one and below the new one include a line like "The portions of this software originating from before the fork are subject to the following license: ". Just fyi :)
